src/engine/runtime.js

```diff
@@ -1012,6 +1013,51 @@ class Runtime extends EventEmitter {
             this._step();
         }, interval);
     }
+
+    // ---------------------------------------------------------------------
+
+    /**
+     * Create event listener for variables.
+     * @param {object} e Blockly "var" event
+     */
+    blocklyListen (e) {
+        // Validate event
+        if (typeof e !== 'object') return;
+        if (typeof e.varId !== 'string') return;
+        const stage = this.getTargetForStage();
+        // Var create/rename/destroy
+        switch (e.type) {
+        case 'var_create':
+            {
+                if (!stage.variables.hasOwnProperty(e.varId)) {
+                    const newVariable = new Variable(e.varId, e.varName, 0,
+                        false);
+                    stage.variables[e.varId] = newVariable;
+                }
+                break;
+            }
+        case 'var_rename':
+            {
+                if (stage.variables.hasOwnProperty(e.varId)) {
+                    const variable = stage.variables[e.varId];
+                    if (variable.id === e.varId) {
+                        variable.name = e.newName;
+                    }
+                }
+                break;
+            }
+        case 'var_delete':
+            {
+                const id = e.varId;
+                if (stage.variables.hasOwnProperty(id)) {
+                    delete stage.variables[id];
+                }
+                break;
+            }
+        }
+    }
+    // ---------------------------------------------------------------------
```

Comment:
For testability and organization, I also think we should push the logic for manipulating the variable list into methods on `Target`, and then call them from the listener rather than having it inline in the listener.